File: LeetCode/0658. Find K Closest Elements/solution.cpp
/**
 * 66 / 66 test cases passed.
 * Runtime: 28 ms
 * Memory Usage: 30.2 MB 
 */
class Solution {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        int right = lower_bound(arr.begin(), arr.end(), x) - arr.begin();
        int left = right - 1;
        while (k --> 0) {
            if (left < 0) {
                ++ right;
            } else if (right >= arr.size()) {
                -- left;
            } else if (x - arr[left] <= arr[right] - x) {
                -- left;
            } else {
                ++ right;
            }
        }
        vector<int> ans(arr.begin() + left + 1, arr.begin() + right);
        return ans;
    }
};

File: LeetCode/0658. Find K Closest Elements/solution.py
"""
66 / 66 test cases passed.
Runtime: 72 ms
Memory Usage: 16.1 MB
"""
class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        right = bisect.bisect_left(arr, x)
        left = right - 1
        while (k := k - 1) >= 0:
            if left < 0:
                right += 1
            elif right >= len(arr) or x - arr[left] <= arr[right] - x:
                left -= 1
            else:
                right += 1
        return arr[left + 1:right]  
